algorithms/src/crh/bhp.rs

```diff
@@ -86,27 +86,25 @@ impl<G: ProjectiveCurve, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> CRH
         }
     }
 
-    fn hash(&self, input: &[u8]) -> Result<Self::Output, CRHError> {
+    fn hash_bits(&self, input: &[bool]) -> Result<Self::Output, CRHError> {
         let eval_time = start_timer!(|| "BoweHopwoodPedersenCRH::hash");
 
-        if (input.len() * 8) > WINDOW_SIZE * NUM_WINDOWS {
+        if input.len() > WINDOW_SIZE * NUM_WINDOWS {
             return Err(CRHError::IncorrectInputLength(input.len(), WINDOW_SIZE, NUM_WINDOWS));
         }
         assert!(WINDOW_SIZE <= MAX_WINDOW_SIZE);
         assert!(NUM_WINDOWS <= MAX_NUM_WINDOWS);
 
         // overzealous but stack allocation
-        let mut buffer = [0u8; MAX_WINDOW_SIZE * MAX_NUM_WINDOWS / 8 + BOWE_HOPWOOD_CHUNK_SIZE + 1];
-        buffer[..input.len()].copy_from_slice(input);
-        let buf_slice = from_bytes_le_to_bits_le(&buffer[..]).collect::<Vec<_>>();
+        let mut buf_slice = input.to_vec();
+        buf_slice.resize(WINDOW_SIZE * NUM_WINDOWS, false);
```

Comment:
Why don't we preserve the existing logic of avoiding allocations?

Reply:
You mean allocate sufficient bits at the beginning (what the `buffer` does) and avoid possible reallocation?

Reply:
in the old code, the buffer allocated sufficient space as a *stack array* (i.e., no heap allocations), and used that everywhere. The new code instead reallocates twice.

Reply:
Revised the implementation back to the old one.

Reply:
Decided to use the new implementation. The old one is somewhat complicated: it has to allocate `MAX_WINDOW_SIZE * MAX_NUM_WINDOWS` instead of `WINDOW_SIZE * NUM_WINDOWS`, so it is much larger. Given that this seems to be adding more complexity while the performance gain might be small, I will go ahead and simplify the logic instead.

Reply:
Ah, change my mind, I will use the old implementation instead, as it does avoid additional allocation.